### ISSUE TITLE
slow syntactic stage

### ISSUE THREAD

**Issue by saidctb** [pyccel/pyccel#481]
**Describe the bug**
due to the use of `_atomic` in here https://github.com/pyccel/pyccel/blob/185f737ef97a57a077a90683757d8ab38b8a7ebc/pyccel/parser/syntactic.py#L725
the parsing becomes slower in the syntactic stage

**Comment by saidctb** [comment]
we also need to check for potential bug when we have nested functions with different return tuple length

**Comment by EmilyBourne** [comment]
Textx is also slowing the code down due to the creation of the meta-model at this line:
```
meta = metamodel_from_file(grammar, debug=debug, classes=hdr_classes)
```
in `headers.py:465(parse)`.
This information can be cached so the model is only created once per python instance. This should increase the speed of our epyccel tests

**Comment by yguclu** [comment]
@EmilyBourne How would you go about caching textX's meta-model?

**Comment by EmilyBourne** [comment]
I'm not certain. Saving it in a global variable would probably be one way

**Comment by yguclu** [comment]
Maybe we could pickle it at installation...

**Comment by EmilyBourne** [comment]
That sounds like a better idea, presumably then it would help for pyccel as well as epyccel

**Comment by EmilyBourne** [comment]
Isn't this PR fixed?

**Comment by yguclu** [comment]
This was fixed by PRs #554 and #665.
